Add curl option to geturl()

dkerr64 opened the pull request:
Fixes issue #232 

Please review / comment @abelbeck @rhansen @SuperSandro2000 

Note, diff makes changes look more substantial than they are.  This is just a block of code inserted within an if {}  and all the existing code is indented inside the else {}.  The indentation causes all the diff.

abelbeck commented:
Does the `header_ok` function need to be short-circuited if `curl` is set ? since  `curl` performs the HTTP functionality and the raw perl network IO HTTP must be done in perl.

dkerr64 commented:
> Does the `header_ok` function need to be short-circuited if `curl` is set ? since `curl` performs the HTTP functionality and the raw perl network IO HTTP must be done in perl.

All of the `header_ok` testing should be removed and placed inside `geturl()` whether it uses curl or perl modules.  I've looked at each place it is checked and none of them actually need to know what is in the header.  `geturl()` should just return the body of the page or an error/undef, header info never needs to be returned.

abelbeck commented:
> All of the header_ok testing should be removed and placed inside geturl() whether it uses curl or perl modules.

Good idea.

rhansen commented on ddclient.in:
```diff
@@ -2020,114 +2022,176 @@ sub geturl {
-    } elsif ($ipversion ne '') {
-        fatal("geturl passed unsupported 'ipversion' value %s", $ipversion);
-    }
+    if (opt('curl')) {
```
Please split this one function into three to make the code easier to read and test, and to make this PR easier to review:

```perl
sub geturl {
    # ...
    my @args = (...);
    my $reply = opt('curl') ? fetch_via_curl(@args) : fetch_via_socket_io(@args);
    # ...
}
```

dkerr64 replied:
Done.

rhansen commented on ddclient.in:
```diff
@@ -2020,114 +2022,176 @@ sub geturl {
+            # don't include ${url} as that might expose login credentials
+            $0 = sprintf("%s - curl (w/headers) sending to %s", $program, "https://${server}");
+            verbose("SENDING:", "curl (w/headers) to %s", "https://${server}");
+            $reply = <<`USE_CURL_WITH_HEADERS`;
+curl -s -X $method --user-agent '${program}/${version}' \\
+              --connect-timeout $timeout --max-time $timeout $curlopt \\
+              ${head_args} \\
+              --data '${data}' \\
+              --url 'https://${server}/${url}' 2>/dev/null
+USE_CURL_WITH_HEADERS
```
Backquotes always make me nervous because it's easy to accidentally introduce a security vulnerability. Even if there is no security risk, unquoted shell metacharacters in the input variables can cause weird problems. I would be more comfortable with a subroutine that acts like backquotes except it takes a list of arguments.

Relevant SO question: https://stackoverflow.com/q/54487856.

dkerr64 replied:
Learning something new every day.  I have created a `curl_cmd()` function and use that to execute curl.  I could only get it to work by passing in a single string with all the curl arguments, trying to pass in each as a separate parameter then passing the array into the `open()` function does not work... got a message back from curl that it did not recognize anything after the first argument..

In other words, this works...
```
sub curl_cmd {
    my $params = shift;
    open(my $fh, "-|", "curl " . $params);
    local $/;
    <$fh>
}
```
But this does not...
```
sub curl_cmd {
    open(my $fh, "-|", "curl" , @_);
    local $/;
    <$fh>
}
```
I suppose I could pass the arguments into `curl_cmd()` individually and then `join(' ',@_)` them all into a single string.

Thoughts?

dkerr64 replied:
Further investigation... both of the options below work.  Its interesting how everything has to be separated in the second version.  Thoughts?
```
#!/usr/bin/perl -w

sub curl_cmd {
    my $params = shift;
    open(my $fh, "-|", "curl " . $params);
    local $/;
    <$fh>
}

sub curl_cmd2 {
    open(my $fh, "-|", "curl",@_);
    local $/;
    <$fh>
}

my $reply = curl_cmd("-s -X GET --url 'http://checkip.dyndns.org/'");
printf("%s\r\n", $reply);
$reply = curl_cmd2("-s","-X","GET","--url","http://checkip.dyndns.org");
printf("%s\r\n", $reply);
exit 0;
```

rhansen replied:
Your `curl_cmd2` function is what we want as it avoids the shell. However, I suggest including the command itself in the list of arguments.

dkerr64 replied:
Currently I have it implemented as `my $reply = qx{ curl --config $tmpfile 2>/dev/null };` which I think is much clearer than using file open method.  Please look at current version of the code.

The file open method is uncommon so most programmers would not immediately recognize what is going on... so unless there is a strong case for it?

rhansen replied:
I left a separate review comment about that line, so I'm going to resolve this conversation.

rhansen commented on ddclient.in:
```diff
@@ -2020,114 +2022,176 @@ sub geturl {
+            my $auth     = '';
+            my $curlopt  = '';
+            my $protocol = 'http';
+            $auth    = "--user '${login}:${password}'" if (($login ne '') && ($password ne ''));
```
Don't pass secrets in command-line arguments—use an environment variable, stdin, or a file instead.

dkerr64 replied:
Not sure how to accomplish this.  Maybe the `--netrc-file` option to curl?  Open to suggestions.

dkerr64 replied:
I think it may be possible to use:
```
xargs curl -s  --url 'http://example.com/' <<< "--user '${login}:${password}'"
```
However this will require opening for both reading and writing... which means `use IPC::Open2`. I have not managed to get curl on its own with `-K-` option to work -- at least not yet.

SuperSandro2000 replied:
Reason behind this is probably that you can read the secrets from other programs like htop.

dkerr64 replied:
> Reason behind this is probably that you can read the secrets from other programs like htop.

Yes exactly.  But piping in from STDIN should close that loophole.  I'm looking into how to get that working.

rhansen replied:
Running
```
xargs curl ... <<<"--user username:password"
```
is just a complicated way of running
```
curl ... --user username:password
```
It still passes the secret via a command-line argument so it will still show up in `ps` and whatnot. Also, `<<<` is a bashism—it's not available in POSIX shell.

The `--config` (`-K`) command-line option is probably what you want, unless curl can take the password via an environment variable.

dkerr64 replied:
Yes, it is now implemented using config file.  See latest code.

dkerr64 commented:
> Pass `--cacert`/`--capath` if `ssl_ca_file`/`ssl_ca_dir` is set.

Yes, and in addition proxy is not implemented either (though arguably that is another thing that can be left to system wide configuration that curl will pick up for us automatically).

abelbeck commented:
Update, 

I meant to keep the -s curl option, fixed:




Lonnie



> On Jul 5, 2020, at 10:49 AM, David Kerr <notifications@github.com> wrote:
> 
> 
> @dkerr64 commented on this pull request.
> 
> In ddclient.in:
> 
> > +              --connect-timeout $timeout --max-time $timeout $curlopt \\
> +              ${head_args} \\
> +              --data '${data}' \\
> +              --url 'https://${server}/${url}' 2>/dev/null
> +USE_CURL_WITH_HEADERS
> +
> +            if (!$reply) {
> +                # don't include ${url} as that might expose login credentials
> +                warning("curl (w/headers) cannot connect to https://%s using IPv%s",${server},$ipversion);
> +            }
> +        } else {
> +            my $timeout  = opt('timeout');
> +            my $auth     = '';
> +            my $curlopt  = '';
> +            my $protocol = 'http';
> +            $auth    = "--user '${login}:${password}'" if (($login ne '') && ($password ne ''));
> 
> Reason behind this is probably that you can read the secrets from other programs like htop.
> 
> Yes exactly. But piping in from STDIN should close that loophole. I'm looking into how to get that working.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

dkerr64 commented:
@rhansen The CI failure(s) I'm getting don't seem to have anything to do with my code?

SuperSandro2000 commented:
> @rhansen The CI failure(s) I'm getting don't seem to have anything to do with my code?

Which do you mean?

dkerr64 commented:
> > @rhansen The CI failure(s) I'm getting don't seem to have anything to do with my code?
> 
> Which do you mean?

Well it seems to be working now.  But there was an error on fedora earlier that looked like it had failed to download some dependency.  If it happens again I'm take a copy of the report.

rhansen commented:
> @rhansen The CI failure(s) I'm getting don't seem to have anything to do with my code?

The Fedora failure is unrelated to your code. I tried to fix it in 5aa4fe71da6f86893633e8a2f5459a9aa40e42aa but I don't think that fixed it entirely (if at all).

rhansen commented on ddclient.in:
```diff
@@ -2146,6 +2154,128 @@ sub geturl {
+    chmod(0600,$tmpfile);
+    print($TFH $params);
+    close($TFH);
+    my $reply = qx{ curl --config $tmpfile 2>/dev/null };
```
Add a test for the curl utility to `configure.ac` and substitute the discovered path (see `subst` in `Makefile.am`).

dkerr64 replied:
Yes I understand.  Will do.

dkerr64 replied:
Done

rhansen commented on ddclient.in:
```diff
@@ -2146,6 +2154,128 @@ sub geturl {
+    # debug("CURL:\r\n%s\r\n",$params); ## Uncomment only if required as may expose credentials
+
+    eval "no warnings 'all'; use File::Spec;";
+    $tmpfile = ((!$@) ? File::Spec->tmpdir() : '/tmp').'/ddclient-';
+    my @chars = ("A".."Z", "a".."z", "0".."9");
+    $tmpfile .= $chars[rand @chars] for 1..10;
+    debug("CURL Tempfile: %s",$tmpfile);
+    open(my $TFH, '>', $tmpfile) or fatal("Falied to open $tmpfile for curl");
```
Use [File::Temp](https://perldoc.perl.org/File/Temp.html).

Better would be to use standard input via something like IPC::Open2.

dkerr64 replied:
I had an implementation that preferred File::Temp and used it only if it existed.  The problem is that it auto deletes the file when you're done with it so when I do the unlink I get warning message that the file is gone already.  But in the case where File::Temp does not exist on the system (rare) you must do the unlink.  I tried everything to suppress that warning but could not get rid of it.  But I can put it back if you want.

IPC::Open2 installed by default on even fewer systems that File::Temp.

rhansen replied:
File::Temp is a core module and exists in Perl v5.10.1, so it is safe to assume it exists.

dkerr64 replied:
How would you like it to fail (or not fail) if File::Temp doesn't exist.
@abelbeck FYI as affects astlinux which does not have this -- would you add it?

SuperSandro2000 replied:
Add it to configure.ac and let them handle it.

dkerr64 replied:
I have added File::Temp.  I'm not sure about handling in configure... we have to cater for the cross-compile use case (building a system on one device for a different target).  This would require something like `--with[out]-file-temp` support which would then have to set some values which would then have to change the code between `ddclient.in` and `ddclient` which I think is bad (as it is, I still don't understand why we need a `.in` version and can't just use `ddclient`). @abelbeck FYI.

rhansen replied:
>  I'm not sure about handling in configure

Just add it to the [list of required Perl modules](https://github.com/ddclient/ddclient/blob/ca1b519444aafe346341800ad80765b72aca0efb/configure.ac#L40-L42) under "Perl modules required to run ddclient."

> we have to cater for the cross-compile use case

`configure.ac` is not set up for that—it assumes that you are running `configure` on the target system itself. If you need cross-build support, please open a new issue. In the meantime, you can patch `configure.ac` to delete the module checks.

> I still don't understand why we need a `.in` version and can't just use `ddclient`

Reason 1: We need a way to substitute user-supplied values that can differ between systems for technical or policy reasons. Examples include:
  * The filename of the `perl` executable.
  * Which directory contains `ddclient.conf`.
  * Where the cache file should be written.

Reason 2: We want to move the documentation source out of `ddclient.in` and into a separate XML or Markdown file so that we can easily build a man page. The documentation will be converted to plain text and substituted back into the resulting `ddclient` script.

Reason 3: In the future we might want to use something like [`pp`](https://metacpan.org/pod/pp) to package the script and its required modules into a single executable.

dkerr64 replied:
Added File::Temp to list of required modules.

rhansen commented on ddclient.in:
```diff
@@ -2146,6 +2154,128 @@ sub geturl {
+    chmod(0600,$tmpfile);
+    print($TFH $params);
+    close($TFH);
+    my $reply = qx{ curl --config $tmpfile 2>/dev/null };
```
Check `$?` to see if curl completed successfully.

dkerr64 replied:
Yes, agree.  Was just discussing that with @abelbeck we need to test for several fail conditions (like invalid certificate, malformed URL, etc.).  Will add that.

dkerr64 replied:
Implemented this with a subset of the more likely return codes issuing a meaningful warning message.  If you think there are others that should be added to this shortlist let me know.

rhansen commented on ddclient.in:
```diff
@@ -2146,6 +2154,128 @@ sub geturl {
+## geturl -- via curl
+######################################################################
+sub curl_cmd {
+    my $params = join("\n",@_);
```
This defeats the purpose of passing the args as a list. (Well, almost: instead of worrying about shell metacharacters we only have to worry about curl config file metacharacters.)

dkerr64 replied:
I removed this and updated the `print` function to take the array directly.

rhansen commented on ddclient.in:
```diff
@@ -2146,6 +2154,128 @@ sub geturl {
+    chmod(0600,$tmpfile);
+    print($TFH $params);
+    close($TFH);
+    my $reply = qx{ curl --config $tmpfile 2>/dev/null };
```
Shell metacharacters in `$tmpfile` need to be quoted. Or use `open -|` with a list.

dkerr64 replied:
Do you mean in the filename itself, or in the contents of the file?  For the contents, we're escaping double quotes before passing to this function.

rhansen replied:
The filename itself.

dkerr64 replied:
We are generating the filename ourself... combination of dir path from the system plus a filename built from `_A-Za-z0-9`.  I don't think there can be any metacharacters that we need to worry about?

dkerr64 replied:
I have single quoted `'$tmpfile'` which should handle it.

rhansen replied:
Unless a malicious actor puts a single quote in `$ENV{TMPDIR}`.

dkerr64 replied:
Should be safe now.  Have hardcoded /tmp and we generate our own filename.

rhansen commented on ddclient.in:
```diff
@@ -2146,6 +2154,128 @@ sub geturl {
+        verbose("SENDING:", "%s", "${server}/${url}");
+    } else {
+        push(@curlopt, "silent");
+        push(@curlopt, "insecure") if ($use_ssl && $is_get_ip);  ## this should really be a user option
```
IIUC, IO::Socket::SSL does validation by default nowadays so this is a security regression.

dkerr64 replied:
Happy to remove this if it is consistent with how you handle it in IO::Socket::SSL implementation.  For context we first implemented curl support several years ago when ddclient did not have its own SSL and when websites may not have properly signed certificates.

rhansen replied:
Yes, remove it.

dkerr64 replied:
removed

rhansen commented on ddclient.in:
```diff
@@ -2146,6 +2154,128 @@ sub geturl {
+        }
+    }
+
+    ## during testing simulate reading the URL
+    if (opt('test')) {
+        my $filename = "$server/$url";
+        $filename =~ s|/|%2F|g;
+        if (opt('exec')) {
+            $reply = save_file("$savedir/$filename", $reply, 'unique');
+        } else {
+            $reply = load_file("$savedir/$filename");
+        }
+    }
+
+    $reply =~ s/\r//g if defined $reply;
```
Move common code to `geturl`.

dkerr64 replied:
Agree.  But remember you were seeking simplicity of code review so I separated out the functions.  I actually think the curl and non-curl should get placed into a single geturl because there are common pieces that wrap both, but I don't want to have to create a new function parameter definitions (all the canonify stuff) that impact both curl and non-curl.

rhansen replied:
I plan on using [URI](https://metacpan.org/pod/URI) to handle URL parsing, and that should minimize argument packing and unpacking between `geturl` and the `fetch_via_*` functions. Stay tuned.

dkerr64 replied:
I'll wait and see what you do... Maybe you would like to implement the `geturl()` wrapper and `fetch_via_socket_io()` functions, then I can sync this pull request to that and insert the `fetch_via_curl()` part?

dkerr64 commented:
> Add unit tests.

Are the geturl unit tests you have been working on not sufficient?  The only place curl_cmd() will be called from is from that and if you insist on separate unit tests then I'm going to move it all back into geturl.  I only moved it out for code clarity.

rhansen commented:
> > Add unit tests.
> 
> Are the geturl unit tests you have been working on not sufficient?

At the very least the `geturl` tests need `curl=true` cases. But not all `geturl` tests are black-box—the tests in PR #240 for example assume IO::Socket and don't apply at all to curl.

SuperSandro2000 commented:
The CI fail should be fixed if you rebase on master.

dkerr64 commented:
> Can you please print the unknown error code so that we really know what it was?

The RC is printed as part of the message.

rhansen commented on ddclient.in:
```diff
@@ -2146,6 +2154,146 @@ sub geturl {
+    my $timeout  = opt('timeout');
+    my @curlopt = ();
+    my @header_lines = ();
+    my $is_get_ip = (((caller(2))[3] // '') eq 'ddclient::get_ip');
```
Use `$params->{ignore_ssl_option}` instead.

dkerr64 replied:
yes, once I catch up with all your changes.

dkerr64 replied:
Done

dkerr64 commented on ddclient.in:
```diff
@@ -2007,7 +2015,7 @@ sub geturl {
     $globals{'fw'} && debug("glo fw = %s", $globals{'fw'});
     ## always omit SSL for connections to local router
-    if ($force_ssl || ($globals{'ssl'} && ((caller(1))[3] // '') ne 'main::get_ip')) {
+    if ($force_ssl || ($globals{'ssl'} && ((caller(2))[3] // '') ne 'ddclient::get_ip')) {
```
I'm trying.  Master is moving so fast it is hard to keep up.  Also I don't have huge success with rebase, when it finds conflicts that I have to manually fix I can end up in a state where rebase hangs and I don't know why.  Not all the time, just sometimes.

dkerr64 commented:
> You can just rebase or cherry pick to keep the history closer together.

I'm trying.... but Master is moving so fast, and some changes are in same areas of code requiring manual intervention which does not always go smoothly.  rebase sometimes hangs on me.

dkerr64 commented on ddclient.in:
```diff
@@ -1995,6 +2001,8 @@ sub geturl {
     my ($sd, $request, $reply);
 
+    use IO::Socket;
+
```
I've reverted the IO::Socket change

rhansen commented:
> > You can just rebase or cherry pick to keep the history closer together.
> 
> I'm trying.... but Master is moving so fast,

Apologies for that.

I find that a good 3-way merge tool is crucial for reliable and easy rebasing. `kdiff3` is the only open source 3-way merge tool I've found that does a decent job, though it is not without its quirks. It will take a while to learn how to use it effectively.

rhansen commented on ddclient.in:
```diff
@@ -2145,6 +2157,157 @@ sub geturl {
+
+    # debug("CURL:\r\n%s\r\n",$params); ## Uncomment only if required as may expose credentials
+
+    eval "no warnings 'all'; use File::Temp;";
+    if (!$@) {
```
I don't like this conditional inclusion because we end up having more code to test and maintain. (It's also not the right way to do conditional inclusion.) If File::Temp isn't available on every system we want to support then we shouldn't use it at all.

Until a compelling use case appears, I'm not willing to jump through hoops to support systems that are missing core modules. Change this to unconditionally import File::Temp (crash out if File::Temp isn't available).

rhansen replied:
> And that is one of the principles you are trying to adhere to—that it will continue to work where it worked before.

No, that's not accurate. If we wanted to maintain compatibility forever we would simply freeze the code and never make any changes. What we're actually trying to do is strike the right balance between code maintainability, features, and support for old systems. See [the Compatibility section of the contributor guide](https://github.com/ddclient/ddclient/blob/master/CONTRIBUTING.md#compatibility) for details.

> the [Astlinux](https://www.astlinux-project.org/) platform does not have it and @abelbeck looked hard at what it would take to add it and found that it drags in a whole pile of other dependencies.

I find that odd, unless Astlinux is missing several core Perl modules. I'd like to know the details.

The cost of working around the lack of File::Temp is too high. Either File::Temp should be added to Astlinux, or this PR should take an approach that doesn't require the creation of a temporary file (e.g., feed the config file to curl via stdin, or use WWW::Curl).

dkerr64 replied:
> The cost of working around the lack of File::Temp is too high. Either File::Temp should be added to Astlinux, or this PR should take an approach that doesn't require the creation of a temporary file (e.g., feed the config file to curl via stdin, or use WWW::Curl).

I'll let @abelbeck address Astlinux.  If he is able to add it then great.  Or he can look at adding IPC::Open2 or WWW::Curl.  I really don't want to get into the arguments -- I just want to get something done and move on.

I have learned a great deal working on this PR and am very happy with how it looks now.  I think we have a great solution for now and should merge and move on.

dkerr64 replied:
Uses File::Temp unconditionally now.

rhansen commented on ddclient.in:
```diff
@@ -2145,6 +2157,157 @@ sub geturl {
+    }
+    chmod(0600,$tmpfile);
+    debug("CURL Tempfile: %s",$tmpfile);
+    local $\ = "\n";  ## Terminate the file,
+    local $, = "\n";  ## and each parameter, with a newline.
```
```suggestion
    $tmpfile->output_field_separator("\n");
    $tmpfile->output_record_separator("\n");
```

rhansen replied:
Did you get those errors with the File::Temp object, or the homemade temporary file handle?

dkerr64 replied:
It doesn't matter how the file is opened I get that warning.  It appears to be a warning because I found documentation for [IO::Handle](https://perldoc.perl.org/IO/Handle.html) that states that you cannot use these methods on a per-file-handle basis and if you try it will just set them globally for all file handles.  To get rid of the warning message we could...
```
IO::Handle->output_record_separator("\n");
IO::Handle->output_field_separator("\n");
```
But then that is global, there is no way to make it local just for this file.  So I prefer to leave it as it is.

rhansen commented on ddclient.in:
```diff
@@ -2145,6 +2157,157 @@ sub geturl {
+    debug("CURL Tempfile: %s",$tmpfile);
+    local $\ = "\n";  ## Terminate the file,
+    local $, = "\n";  ## and each parameter, with a newline.
+    print($tmpfile @params);
```
Metacharacters in the entries of `@params` need to be quoted.

rhansen replied:
My concern isn't with the location of the quoting, but with the lack of quoting. `fetch_via_curl` doesn't sanitize user input by quoting characters that are special in a curl config file (double quotes, newlines, spaces, etc.).

dkerr64 replied:
> My concern isn't with the location of the quoting, but with the lack of quoting. `fetch_via_curl` doesn't sanitize user input by quoting characters that are special in a curl config file (double quotes, newlines, spaces, etc.).

I have added `quotemeta()` for the file path when not using `File::Temp` this should protect for when the system ENV is used.  I have also escaped those characters recognized by curl (see --config in curl [manpage](https://curl.haxx.se/docs/manpage.html))

Let me know if you see anything else that should be done. 

Thanks

dkerr64 replied:
I think this is all resolved now.

rhansen commented on ddclient.in:
```diff
@@ -2145,6 +2157,157 @@ sub geturl {
+    ## Make sure to delete temp file on Ctrl-C or SIGTERM
+    local $SIG{'INT'} = sub { unlink($tmpfile); exit 2 };
+    local $SIG{'TERM'} = sub { unlink($tmpfile); exit 15 };
+    my $reply = qx{ /usr/bin/curl --config '$tmpfile' 2>/dev/null; };
```
What does `curl` output? Is it the full HTTP response including the status line and headers? The return value should match what `fetch_via_socket_io` returns. We can change `fetch_via_socket_io` if that is easier than changing `fetch_via_curl`.

rhansen replied:
I'm still working on #170. It's tedious to audit every function to make sure I'm not breaking anything, so it'll be a while.

Let's leave this unresolved until either I finish #170 or this is changed to return the full response.

dkerr64 replied:
The vast majority of places where it is used it just checks `!header_ok()` and move on if it comes back not ok... which means anything other than an http 200.  Those are easy to handle and in almost every case the header_ok function call would not even be executed (because geturl would have returned '' and that is checked before header).

In the few places that get a positive reply the first thing they do is strip off the header, so it is thrown away.

My point is, it's not as hard as you think.

dkerr64 replied:
Curl version now returns HTTP headers for compatibility.  We can address in #170 which you can assign to me if you want after this is merged.

rhansen commented on ddclient.in:
```diff
@@ -811,6 +814,7 @@ sub main {
     $SIG{'HUP'}  = sub { $caught_hup  = 1; };
     $SIG{'TERM'} = sub { $caught_term = 1; };
     $SIG{'INT'}  = sub { $caught_int  = 1; };
+
```
Undo this line add. (It might be better to have this blank line, but this PR isn't the right place to add it.)

dkerr64 commented:
So I decided to look at the unit tests.  Boy, a lot of dependencies needed to get installed before I could run the geturl_connectivity tests.  Interesting results, first with me enabling curl...
```
PASS: t/geturl_connectivity.pl 1 - IPv* client to http://127.0.0.1:40443 (-noipv6)
PASS: t/geturl_connectivity.pl 2 - IPv* client to https://127.0.0.1:46465 (-noipv6)
PASS: t/geturl_connectivity.pl 3 - IPv4 client to http://127.0.0.1:40443 (-noipv6)
PASS: t/geturl_connectivity.pl 4 - IPv4 client to https://127.0.0.1:46465 (-noipv6)
PASS: t/geturl_connectivity.pl 5 - IPv6 client to http://[::1]:37885 (-noipv6)
PASS: t/geturl_connectivity.pl 6 - IPv6 client to https://[::1]:42577 (-noipv6)
PASS: t/geturl_connectivity.pl 7 - IPv* client to http://127.0.0.1:40443 (-ipv6)
PASS: t/geturl_connectivity.pl 8 - IPv* client to https://127.0.0.1:46465 (-ipv6)
PASS: t/geturl_connectivity.pl 9 - IPv* client to http://[::1]:37885 (-ipv6)
PASS: t/geturl_connectivity.pl 10 - IPv* client to https://[::1]:42577 (-ipv6)
PASS: t/geturl_connectivity.pl 11 - IPv4 client to http://127.0.0.1:40443 (-ipv6)
PASS: t/geturl_connectivity.pl 12 - IPv4 client to https://127.0.0.1:46465 (-ipv6)
PASS: t/geturl_connectivity.pl 13 - IPv6 client to http://[::1]:37885 (-ipv6)
PASS: t/geturl_connectivity.pl 14 - IPv6 client to https://[::1]:42577 (-ipv6)
PASS: t/geturl_connectivity.pl 15 - no (unexpected) warnings (via done_testing)
```
Second without curl...
```
PASS: t/geturl_connectivity.pl 1 - IPv* client to http://127.0.0.1:41395 (-noipv6)
PASS: t/geturl_connectivity.pl 2 - IPv* client to https://127.0.0.1:33601 (-noipv6)
PASS: t/geturl_connectivity.pl 3 - IPv4 client to http://127.0.0.1:41395 (-noipv6)
PASS: t/geturl_connectivity.pl 4 - IPv4 client to https://127.0.0.1:33601 (-noipv6)
PASS: t/geturl_connectivity.pl 5 - no (unexpected) warnings (via END block)
ERROR: t/geturl_connectivity.pl - missing test plan
ERROR: t/geturl_connectivity.pl - exited with status 1
```
Oops... failed as soon as it got to IPv6 tests.  This is on Ubuntu 20.04 in VMware which does not have IPv6 connection to the internet but does have IPv6 enabled and a link local address on the network interface, so it should have worked just fine going to loopback. 

Without intending to be condescending, but this is exactly why we need curl as an option.  Of course there could be a simple explanation and an easy fix at my end.

@abelbeck FYI

rhansen commented:
> failed as soon as it got to IPv6 tests. This is on Ubuntu 20.04 in VMware which does not have IPv6 connection to the internet but does have IPv6 enabled and a link local address on the network interface

Does current `master` pass? If not, please open a new issue and paste the relevant lines from `t/geturl_connectivity.log`.

If `master` does pass, then paste the relevant log lines here and I can take a look.

> this is exactly why we need curl as an option

I suspect the failures are unrelated to curl. The `t/geturl_connectivity.pl` test sets up a few web servers listening on 127.0.0.1 and ::1 so that there is something for `geturl` to talk to. The logic to set up those web servers is unrelated to `geturl`, and highly sensitive to platform-specific issues. It's possible those web servers are having problems.

SuperSandro2000 commented on ddclient.in:
```diff
@@ -2145,6 +2156,157 @@ sub geturl {
+
+    ## Make sure to delete temp file on Ctrl-C or SIGTERM
+    local $SIG{'INT'} = sub { unlink($tmpfile); exit 2 };
+    local $SIG{'TERM'} = sub { unlink($tmpfile); exit 15 };
```
Took me at least 10 minutes to find this comment. GitHub grouped them all together in the mail and I couldn't open this specific...

The file definitely needs to be cleaned up and Multiple Ctrl+C is usually annoying...

dkerr64 replied:
Current code should handle this properly, no multiple Ctrl-C required.

rhansen replied:
> Took me at least 10 minutes to find this comment.

GitHub's review UI is the worst. I wonder if GitLab has a better review UI.

> Thoughts?

It's hard to get signals right, and the lack of good documentation doesn't help.

For now I wouldn't worry about cleaning up the temporary files if there's a signal. It's not ideal, but the consequences are minor: ddclient is likely to only be terminated once per system reboot, and reboots should clean out stale temporary files. And the probability that ddclient is terminated while curl is running is quite low. Incorrectly handling signals can cause more problems than leaving the occasional temporary file.

Eventually I think we should redo signal handling. Maybe something like [sigtrap](https://perldoc.perl.org/sigtrap.html) combined with registered closures that are executed in [`END` blocks](https://perldoc.perl.org/perlmod.html#BEGIN%2c-UNITCHECK%2c-CHECK%2c-INIT-and-END).  There are also a bunch of scope-based cleanup modules that we might be able to use:
  * [Scope::Cleanup](https://metacpan.org/pod/Scope::Cleanup)
  * [Scope::OnExit](https://metacpan.org/pod/Scope::OnExit)
  * [Scope::Guard](https://metacpan.org/pod/Scope::Guard)
  * [Guard](https://metacpan.org/pod/Guard)
  * [Sub::ScopeFinalizer](https://metacpan.org/pod/Sub::ScopeFinalizer)
  * [Hook::Scope](https://metacpan.org/pod/Hook::Scope)
  * [B::Hooks::EndOfScope](https://metacpan.org/pod/B::Hooks::EndOfScope)
  * [End](https://metacpan.org/pod/End)
  * [Object::Destroyer](https://metacpan.org/pod/Object::Destroyer)
  * [Perl::AtEndOfScope](https://metacpan.org/pod/Perl::AtEndOfScope)
  * [ReleaseAction](https://metacpan.org/pod/ReleaseAction)
  * [Scope::local_OnExit](https://metacpan.org/pod/Scope::local_OnExit)
  * [Value::Canary](https://metacpan.org/pod/Value::Canary)
  * probably several others

I don't know if any are suitable. I suspect most of them are not packaged, and many are C-based. It probably wouldn't be hard to write our own via object destructor.

dkerr64 replied:
I don't like the idea of knowingly leaving the temp file lying around.  Of all the security issues we have been discussing this is as bad as they come as we knowingly leave an attack vector which could expose credentials.  The current code installs signal handlers, it is very easy to add local signal handlers just for this subroutine and mimic exactly what the current ones do while also deleting the temp file, and if not running as a daemon also exit.  Note that this has nothing to do with how the temp file is created... File::Temp has the same exposure.

rhansen replied:
> I don't like the idea of knowingly leaving the temp file lying around. Of all the security issues we have been discussing this is as bad as they come as we knowingly leave an attack vector which could expose credentials.

How is it a problem? The file is owned by the ddclient user with 0600 permissions. That's the same as `/etc/ddclient.conf`, which also has the sensitive credentials.

> The current code installs signal handlers, it is very easy to add local signal handlers just for this subroutine and mimic exactly what the current ones do while also deleting the temp file, and if not running as a daemon also exit.

I just realized that we don't need to do anything special for signals—the existing signal handlers are sufficient. If a signal arrives while curl is running the existing signal handler will set the appropriate `$caught_*` variable, then normal execution will resume. Curl will have a chance to complete, as will `geturl`.

dkerr64 replied:
> I just realized that we don't need to do anything special for signals—the existing signal handlers are sufficient. If a signal arrives while curl is running the existing signal handler will set the appropriate `$caught_*` variable, then normal execution will resume. Curl will have a chance to complete, as will `geturl`.

Well that is not how it is working. SIGINT is interrupting curl and no file deletion takes place.  If running in a daemon then setting of `$caught_int` or `$caught_term` breaks out of the while loop and program ends.

rhansen replied:
Huh, that's unexpected. Can you tell me how you tested it so that I can reproduce?

dkerr64 replied:
Ah, so that's interesting.  I have found the reason why.  To reproduce do the following...
`ddclient -query`
Then hit Ctrl-C as it is displaying the IP addresses.  Now the reason why is because the call to `test_possible_ip` is done immediately before the signal handlers are set.  So we're both right.  If I move the test_possible_ip call to after the signal handlers then you are right, Ctrl-C does nothing (at least while using -query flag).

So in fact adding my signal handlers changes behavior for the non-deamon state allowing the program to exit.  I need to do further testing in the non -query states.

dkerr64 replied:
Signal handlers have been removed.

abelbeck commented:
Hi @rhansen 
> I find that odd, unless Astlinux is missing several core Perl modules. I'd like to know the details.

I'd love to share.  AstLinux overlays [perl-cross](https://arsv.github.io/perl-cross/download.html) over perl so we can cross-compile it.  By default we disable all modules, and selectively only build crucial modules, mostly related to the Asterisk PBX and ddclient. (The AstLinux compressed image is less than 60 MB)

This set of modules runs classic ddclient (and Asterisk, etc.):
```
PERL_MODULES = constant version base fields
PERL_MODULES += Carp Errno Fcntl PathTools POSIX Digest Socket IO XSLoader Exporter B File-Find JSON-PP
PERL_MODULES += Digest/MD5 Digest/SHA Getopt/Long Time/Local File/Glob Sys/Hostname
```
In order to use `File::Temp` we would need to add (through trial and error):
```
PERL_MODULES += File/Temp File/Path Scalar-List-Utils parent
```
But the fundamental question, is any system without `/tmp/` `ddclient` worthy ?  If yes, I'd like know what that would be.

As @rhansen mentioned `File::Temp` uses `File::Spec->tmpdir()` which can source an environmental variable which is passed to the `curl` command line.

IMHO, this is secure and simple, and what we now do in our `ddclient-curl`.
```
        my @chars = ('A'..'Z', 'a'..'z', '0'..'9');
        my $curlopt_file = '/tmp/ddclient_curlopt_';
        $curlopt_file .= $chars[rand @chars] for 1..10;
        open(my $cfh, '>', $curlopt_file) or fatal("Falied to open $curlopt_file for curl");
...
        $reply = qx{ /usr/bin/curl --config '$curlopt_file' 2>/dev/null ; rm -f '$curlopt_file' };
```

rhansen commented:
> In order to use `File::Temp` we would need to add (through trial and error):
> 
> ```
> PERL_MODULES += File/Temp File/Path Scalar-List-Utils parent
> ```

What is the cost of those additional modules (in terms of image size or other metrics you consider important)? Is that extra cost high enough that it would be better for you to maintain a separate fork of ddclient?

> 
> But the fundamental question, is any system without `/tmp/` `ddclient` worthy ? If yes, I'd like know what that would be.

There are multiple distinct issues that I think are getting mixed up:
  * Issue 1: Defense against special characters in the temporary file's filename.
  * Issue 2: Creating the temporary file and managing its lifetime, regardless of its filename.
  * Issue 3: Defense against special characters in the curl config values that are written to the temporary file.

I was referring to issue 2 when I was saying that we should use File::Temp. Creating a homemade File::Temp alternative imposes too much of a maintenance burden on us relative to its value.

(Regarding issue 1: Hard-coding the filename to something like `/tmp/ddclient_XXXXX` would likely be an acceptable way to defend against special characters in the filename. I'm unaware of any Unix-ish systems that don't have a usable `/tmp`, though maybe there are some weird home routers.)

abelbeck commented:
>>    In order to use File::Temp we would need to add (through trial and error):

>> PERL_MODULES += File/Temp File/Path Scalar-List-Utils parent

>What is the cost of those additional modules (in terms of image size or other metrics you consider important)? Is that extra cost high enough that it would be better for you to maintain a separate fork of ddclient?

@rhansen : The cost adds about 256KB uncompressed (notable, but not a huge deal), If the only added perl module was `File::Temp` we may add it.  But I don't like the security aspects of `File::Temp` for this use case. "Issue 1"

"Issue 2", The lifetime of this temp file is typically milliseconds and is called synchronously.  I like the idea of the `curl` subshell removing the temp file after `curl` is finished with it (above). The only time multiple `curl` calls could exist is when `ddclient` is issued a `SIGTERM` and a new `ddclient` is started before the original daemon loop has completed.

>  I'm unaware of any Unix-ish systems that don't have a usable /tmp, though maybe there are some weird home routers.)

I'm not aware of any either, I suppose the `/tmp/` could be autoconf'ed if needed in the future.

rhansen commented:
> But I don't like the security aspects of `File::Temp` for this use case. "Issue 1"

Why not?  Doesn't the following meet the requirements?

```perl
my $tmpfile = File::Temp->new(DIR => "/tmp", TEMPLATE => "ddclient_XXXXXX");
```

dkerr64 commented:
To get past the arguments, I have removed all the code to test for `File::Temp` and use alternate temp file creation and just assume it exists.  If this is a problem for any specific platform we can open a new issue and tackle it then.

dkerr64 commented:
Stripping out the headers in `geturl()` causes ssl unit tests to fail?  Why are unit tests dependent on headers when none of the code that uses `geturl()` is dependent on them?

dkerr64 commented:
So I reverted removal of header_ok function and added HTTP response headers to the curl version for compatibility.  But I really think it is unnecessary to return the HTTP response to each server protocol subroutine as there is no requirement for it.  At least my commit/revert documents the changes that I think would be needed.

dkerr64 commented:
@rhansen I have added curl to the test suite and find path in configure/make.  Question for you... in the test suite I test for existence of curl and skip if it is not there.  But it is still hard coded to `/usr/bin/curl` in the test program.  We should update the makefile to do the substitution in that file as well as ddclient.in.  How would you like that done?  I don't see any other test files with the same issue yet.

rhansen commented:
> But it is still hard coded to `/usr/bin/curl` in the test program. We should update the makefile to do the substitution in that file as well as ddclient.in. How would you like that done? I don't see any other test files with the same issue yet.

Follow what `t/version.pl.in` does. See `git grep t/version.pl`.

rhansen commented on t/geturl_connectivity.pl:
```diff
@@ -84,8 +87,16 @@ sub run_httpd {
+        $ddclient::globals{'curl'} = 0;
         my $got = ddclient::geturl({ url => $uri, ipversion => $tc->{client_ipv} });
         isnt($got // '', '', $name);
+        SKIP: {
+            skip("Curl not available on this system", 1) if !$has_curl;
+            $name = $name . " (curl)";
+            $ddclient::globals{'curl'} = 1;
+            $got = ddclient::geturl({ url => $uri, ipversion => $tc->{client_ipv} });
+            isnt($got // '', '', $name);
+        }
```
Add a new `curl` key to the entries in `@test_cases` instead of adding a new test to the loop body. Rationale:
  * It makes it easier to mark a test as TODO if necessary.
  * curl tests should not be skipped if IO::Socket::INET6 is not available.
  * Some cases that don't make sense for IO::Socket do make sense for curl.

dkerr64 replied:
Yes, well.  That was more work!  I'll look into this.

dkerr64 replied:
Done.  @rhansen Please check that I got the skips right.

rhansen commented on t/geturl_ssl.pl:
```diff
@@ -57,6 +57,7 @@
         want_args => {
             PeerAddr => 'hostname',
             PeerPort => '443',
+            SSL_verify_mode => 1,
```
Use `SSL_VERIFY_PEER`, not its value.

dkerr64 replied:
It occurred to me after I committed this that an alternative would be to only set `SSL_verify_mode` in ddclient.in if the value was not `SSL_VERIFY_PEER`.... as that is the default.  That would mean no need to change the test cases.  @rhansen can you tell me which you prefer... update test cases to use SSL_VERIFY_PEER or change ddclient to only set the value if non-default (and remove all these lines from the test cases).

rhansen replied:
Either way is fine with me.

IIUC, SSL_VERIFY_PEER did not used to be the default, and that wasn't very long ago, so passing SSL_VERIFY_PEER will give users with old systems better security. I like that. But it does add a bit of complexity, so `¯\_(ツ)_/¯`.

dkerr64 replied:
I'll work on updating the test case, as yes -- it makes most sense to explicitly set it one way or the other than make any assumptions on what the default is.

dkerr64 replied:
Done

rhansen commented on ddclient.in:
```diff
@@ -2155,6 +2168,171 @@ sub geturl {
+    close($tfh);
+
+    ## Make sure to delete temp file on Ctrl-C or SIGTERM
+    local $SIG{'HUP'} = sub { unlink($tmpfile); $caught_hup = 1; exit 1 if !$daemon; };
+    local $SIG{'INT'} = sub { unlink($tmpfile); $caught_term = 1; exit 2 if !$daemon; };
+    local $SIG{'TERM'} = sub { unlink($tmpfile); $caught_int = 1; exit 15 if !$daemon; };
```
Delete this signal handling. It only makes a difference if curl is running outside the main loop and a signal arrives, which never happens in normal operation. Even if it did, the readability and maintenance cost is much worse than the occasional lingering temp file.

dkerr64 replied:
Done

dkerr64 commented:
> > But it is still hard coded to `/usr/bin/curl` in the test program. We should update the makefile to do the substitution in that file as well as ddclient.in. How would you like that done? I don't see any other test files with the same issue yet.
> 
> Follow what `t/version.pl.in` does. See `git grep t/version.pl`.

Just so I am clear... you are asking that I delete `t/geturl_connectivity.pl` and replace it with `t/geturl_connectivity.pl.in` and update the configure and make file accordingly, just like `t/version.pl` is built from `t/version.pl.in`

dkerr64 commented on ChangeLog.md:
```diff
@@ -19,6 +19,13 @@ repository history](https://github.com/ddclient/ddclient/commits/master).
     certificates.
+  * Added option `-curl` to access network with system Curl command instead
+    of the Perl built-in IO::Socket classes.
+  * Added option `-{no}ssl-validate` to provide option to disable SSL
+    certificate validation.  Note that `-nossl-validate` only applies for
+    network access when obtaining an IP address with `use=web` or `use=fw`
+    (any firewall).  Network access to Dynamic DNS servers to set or retrieve
+    IP address will always require certificate validation.
```
It is independently controllable -- at least to the extent that this is a service level option, not global.  And a service can use either `use=web` or `use=fw` not both so the `ssl-validate` applies only to one or the other depending on which of the `use=` is set.  I think it unnecessary to create two separate settings.

rhansen replied:
Good points. However, I am concerned about user confusion from the generic name—the name makes it sound like it applies everywhere, not just `fw` and `web`.

dkerr64 replied:
Maybe the ssl-validate stuff should move to a separate pull request after base curl support is merged.

rhansen replied:
:+1: or before base curl support is merged

dkerr64 replied:
That feels like to much work.  I have separated it into the two options as requested.

dkerr64 commented on ddclient.in:
```diff
@@ -2155,6 +2168,171 @@ sub geturl {
+
+    ## Use system File::Temp to generate filename...
+    use File::Temp;
+    $tfh = File::Temp->new( TMPDIR => 1,  ## in system temp dir
```
So if we're going to just assume `/tmp` then why bother using `File::Temp` in the first place?  FYI on my MacOS `TMPDIR=/var/folders/cq/180yzlm13g5c_66tv8gky5pw0000gn/T/` so there are systems that provide something other than `/tmp` (though MacOS has this as well).

rhansen replied:
> why bother using `File::Temp` in the first place?

Because File::Temp does other stuff too, such as `XXXXXX` substitution and automatic file deletion at scope exit.

> there are systems that provide something other than `/tmp` (though MacOS has this as well).

As long as everyone supports `/tmp` then using `/tmp` avoids the need to escape shell metacharacters in `$ENV{TMPDIR}`, I'm OK with using `TMPDIR => 1` instead of `DIR => '/tmp'` as long as you properly escape special characters in the resulting pathname.

dkerr64 replied:
I have hard coded it to /tmp/

dkerr64 commented on ddclient.in:
```diff
@@ -2155,6 +2168,171 @@ sub geturl {
+    ## Use system File::Temp to generate filename...
+    use File::Temp;
+    $tfh = File::Temp->new( TMPDIR => 1,  ## in system temp dir
+                            UNLINK => 0,  ## we will delete it ourselves
```
I was getting warning messages that the file had already been deleted on exiting the function that I could not suppress.  This was potentially relayed to handling SIGINT etc. and because I had alternatives to File::Temp.  But as we're going to remove all that then we can probably revert back to auto deletion -- I will test and update change it if there are no problems.

dkerr64 replied:
I have updated to use the built-in feature of File::Temp to automatically delete it.

dkerr64 commented on configure.ac:
```diff
@@ -27,6 +27,11 @@ AC_PROG_MKDIR_P
 AS_IF([test -z "${FIND}"], [AC_MSG_ERROR(['find' utility not found])])
 
+# Look for location of curl on this system in case it is something
+# other than /usr/bin/curl
```
Removed

dkerr64 commented on configure.ac:
```diff
@@ -27,6 +27,11 @@ AC_PROG_MKDIR_P
+# Look for location of curl on this system in case it is something
+# other than /usr/bin/curl
+AC_PATH_PROG([CURL], [curl], "/usr/bin/curl")
+AC_SUBST([CURL])
```
removed

dkerr64 commented on ddclient.in:
```diff
@@ -1932,7 +1937,7 @@ On Red Hat, the package perl-IO-Socket-SSL must be installed.
 EOM
     }
-    import IO::Socket::SSL;
+    use IO::Socket::SSL qw( SSL_VERIFY_NONE SSL_VERIFY_PEER );
```
That change doesn't fix the problem, generates an error during compile.
```
Bareword "SSL_VERIFY_PEER" not allowed while "strict subs" in use at ddclient.in line 2088.
Bareword "SSL_VERIFY_NONE" not allowed while "strict subs" in use at ddclient.in line 2088.
```
The only way I could solve that was move from `import` to `use`.  Do you have another suggestion?

dkerr64 replied:
I think I solved this one.  Please take a look @rhansen

dkerr64 commented on ddclient.in:
```diff
@@ -2155,6 +2168,171 @@ sub geturl {
+
+    return '' if (scalar(@params) == 0); ## no parameters provided
+
+    # debug("CURL: %s\r\n",join("\r\n      ",@params)); ## Uncomment only if required as may expose credentials
```
It is NOT okay to include credentials in debug, at least not in my book.  It is very common for people to copy/paste their debug output in bug reports and not all credentials are obvious that that is what they are.  So I do not think we should uncomment this. 

We could delete it, but do you really feel that strongly?  I have found it very useful while developing to be able to easily uncomment this.

rhansen replied:
> It is NOT okay to include credentials in debug

There's a trade-off. It's not always possible to censor/omit sensitive information, but we do need debug information on occasion. I think we should address this by using `verbose()` more often so that `verbose=yes debug=no` logs contain enough information most of the time.

> We could delete it, but do you really feel that strongly?

Yes. Commented code tends to rot over time, it impairs readability, and it makes the surrounding code harder to maintain. If the code is frequently used then it should instead be guarded by a flag/option (such as `-debug`). If it's not frequently used then it's no big deal to recreate it if/when needed.

dkerr64 replied:
I will delete the line

rhansen commented:
> > > But it is still hard coded to `/usr/bin/curl` in the test program. We should update the makefile to do the substitution in that file as well as ddclient.in. How would you like that done? I don't see any other test files with the same issue yet.
> > 
> > 
> > Follow what `t/version.pl.in` does. See `git grep t/version.pl`.
> 
> Just so I am clear... you are asking that I delete `t/geturl_connectivity.pl` and replace it with `t/geturl_connectivity.pl.in` and update the configure and make file accordingly, just like `t/version.pl` is built from `t/version.pl.in`

Yes.

dkerr64 commented on ddclient.in:
```diff
@@ -2155,6 +2168,171 @@ sub geturl {
+    my @params = @_;
+    my $tmpfile;
+    my $tfh;
+    my $system_curl = subst_var('@syscurlcmd@', '/usr/bin/curl');
```
No, there should be a default just like @sysconfdir@ and @localstatedir@ so that it does not error out.

rhansen replied:
Sorry, I wasn't clear. What I mean is this:

```perl
    my $system_curl = subst_var('@CURL@', '/usr/bin/curl');
    fatal("curl not found") if ($system_curl eq '');
```

This way, if `@CURL@` is replaced with the empty string it won't try to execute the empty string.

And now that I think about it more, perhaps we should default to a non-absolute `curl` pathname in `configure.ac` so that it can be found via a `PATH` search:

```autoconf
AC_PATH_PROG([CURL], [curl], [curl])
```

dkerr64 commented:
> > > > But it is still hard coded to `/usr/bin/curl` in the test program. We should update the makefile to do the substitution in that file as well as ddclient.in. How would you like that done? I don't see any other test files with the same issue yet.
> > > 
> > > 
> > > Follow what `t/version.pl.in` does. See `git grep t/version.pl`.
> > 
> > 
> > Just so I am clear... you are asking that I delete `t/geturl_connectivity.pl` and replace it with `t/geturl_connectivity.pl.in` and update the configure and make file accordingly, just like `t/version.pl` is built from `t/version.pl.in`
> 
> Yes.

Done

rhansen commented on ddclient.in:
```diff
@@ -2170,6 +2183,163 @@ sub geturl {
+
+    ## Use system File::Temp to generate filename...
+    use File::Temp;
+    $tfh = File::Temp->new( TMPDIR => "/tmp/",
```
`TMPDIR` should be `DIR`:

```suggestion
    $tfh = File::Temp->new(DIR => '/tmp',
```

rhansen commented on ddclient.in:
```diff
@@ -2170,6 +2183,163 @@ sub geturl {
+
+    ## Use system File::Temp to generate filename...
+    use File::Temp;
+    $tfh = File::Temp->new( TMPDIR => "/tmp/",
```
Add a comment explaining why we're using `DIR` instead of `TMPDIR`.

dkerr64 replied:
Done

rhansen commented on t/geturl_connectivity.pl.in:
```diff
@@ -55,35 +58,52 @@ sub run_httpd {
-    { ipv6_opt => 1, client_ipv => '4', server_ipv => '4', ssl => 1 },
-    { ipv6_opt => 1, client_ipv => '6', server_ipv => '6', ssl => 0 },
-    { ipv6_opt => 1, client_ipv => '6', server_ipv => '6', ssl => 1 },
+    { ipv6_opt => 0, client_ipv => '',  server_ipv => '4', ssl => 0, curl => 0 },
+    { ipv6_opt => 0, client_ipv => '',  server_ipv => '4', ssl => 1, curl => 0 },
+    { ipv6_opt => 0, client_ipv => '4', server_ipv => '4', ssl => 0, curl => 0 },
+    { ipv6_opt => 0, client_ipv => '4', server_ipv => '4', ssl => 1, curl => 0 },
+    { ipv6_opt => 0, client_ipv => '6', server_ipv => '6', ssl => 0, curl => 0 },
+    { ipv6_opt => 0, client_ipv => '6', server_ipv => '6', ssl => 1, curl => 0 },
+    { ipv6_opt => 1, client_ipv => '',  server_ipv => '4', ssl => 0, curl => 0 },
+    { ipv6_opt => 1, client_ipv => '',  server_ipv => '4', ssl => 1, curl => 0 },
+    { ipv6_opt => 1, client_ipv => '',  server_ipv => '6', ssl => 0, curl => 0 },
+    { ipv6_opt => 1, client_ipv => '',  server_ipv => '6', ssl => 1, curl => 0 },
+    { ipv6_opt => 1, client_ipv => '4', server_ipv => '4', ssl => 0, curl => 0 },
+    { ipv6_opt => 1, client_ipv => '4', server_ipv => '4', ssl => 1, curl => 0 },
+    { ipv6_opt => 1, client_ipv => '6', server_ipv => '6', ssl => 0, curl => 0 },
+    { ipv6_opt => 1, client_ipv => '6', server_ipv => '6', ssl => 1, curl => 0 },
+    { ipv6_opt => 0, client_ipv => '',  server_ipv => '4', ssl => 0, curl => 1 },
+    { ipv6_opt => 0, client_ipv => '',  server_ipv => '4', ssl => 1, curl => 1 },
+    { ipv6_opt => 0, client_ipv => '4', server_ipv => '4', ssl => 0, curl => 1 },
+    { ipv6_opt => 0, client_ipv => '4', server_ipv => '4', ssl => 1, curl => 1 },
+    { ipv6_opt => 0, client_ipv => '6', server_ipv => '6', ssl => 0, curl => 1 },
+    { ipv6_opt => 0, client_ipv => '6', server_ipv => '6', ssl => 1, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '',  server_ipv => '4', ssl => 0, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '',  server_ipv => '4', ssl => 1, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '',  server_ipv => '6', ssl => 0, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '',  server_ipv => '6', ssl => 1, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '4', server_ipv => '4', ssl => 0, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '4', server_ipv => '4', ssl => 1, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '6', server_ipv => '6', ssl => 0, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '6', server_ipv => '6', ssl => 1, curl => 1 },
```
Style nit: Move `curl => 1|0` to the beginning, like this:

```perl
    { curl => 1, ipv6_opt => 0, client_ipv => '',  server_ipv => '4', ssl => 0 },
```

dkerr64 replied:
Done

rhansen commented on t/geturl_connectivity.pl.in:
```diff
@@ -55,35 +58,52 @@ sub run_httpd {
+    { ipv6_opt => 0, client_ipv => '4', server_ipv => '4', ssl => 1, curl => 1 },
+    { ipv6_opt => 0, client_ipv => '6', server_ipv => '6', ssl => 0, curl => 1 },
+    { ipv6_opt => 0, client_ipv => '6', server_ipv => '6', ssl => 1, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '',  server_ipv => '4', ssl => 0, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '',  server_ipv => '4', ssl => 1, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '',  server_ipv => '6', ssl => 0, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '',  server_ipv => '6', ssl => 1, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '4', server_ipv => '4', ssl => 0, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '4', server_ipv => '4', ssl => 1, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '6', server_ipv => '6', ssl => 0, curl => 1 },
+    { ipv6_opt => 1, client_ipv => '6', server_ipv => '6', ssl => 1, curl => 1 },
```
`fetch_via_curl` doesn't care about `$globals{'ipv6'}`, so omit the `ipv6_opt => 1` cases.

dkerr64 replied:
Yes, true.  And neither should `fetch_via_socket_io` now that ip version is passed in as a parameter.  But that change should wait until we progress further with IPv6 support -- need my two PRs merged before that can commence.

dkerr64 replied:
done

rhansen commented on t/geturl_connectivity.pl.in:
```diff
@@ -55,35 +58,52 @@ sub run_httpd {
+                           $tc->{client_ipv} || '*', $uri, $tc->{ipv6_opt} ? '' : 'no',
+                           $tc->{curl} ? ' (curl)' : '');
         $ddclient::globals{'ipv6'} = $tc->{ipv6_opt};
+        $ddclient::globals{'curl'} = $tc->{curl};;
```
Superfluous semicolon:
```suggestion
        $ddclient::globals{'curl'} = $tc->{curl};
```

rhansen commented on t/geturl_connectivity.pl.in:
```diff
@@ -55,35 +58,52 @@ sub run_httpd {
+        my $name = sprintf("IPv%s client to %s (-%sipv6)%s",
+                           $tc->{client_ipv} || '*', $uri, $tc->{ipv6_opt} ? '' : 'no',
+                           $tc->{curl} ? ' (curl)' : '');
         $ddclient::globals{'ipv6'} = $tc->{ipv6_opt};
+        $ddclient::globals{'curl'} = $tc->{curl};;
```
If you default these to 0:

```suggestion
        $ddclient::globals{'ipv6'} = $tc->{ipv6_opt} // 0;
        $ddclient::globals{'curl'} = $tc->{curl} // 0;
```

then you can simplify the test cases:

```perl
    # Fetch via IO::Socket::INET or IO::Socket::SSL
    { ipv6_opt => 0, ... },
    ...
    # Fetch via IO::Socket::INET6 or IO::Socket::SSL
    { ipv6_opt => 1, ... },
    ...
    # Fetch via curl
    { curl => 1,     ... },
    ...
```

rhansen replied:
Actually you would have to set the default earlier on so that the `skip` checks would work properly:

```perl
for my $tc (@test_cases) {
    $tc->{ipv6_opt} //= 0;
    $tc->{curl} //= 0;
    SKIP: {
        # ...
```

dkerr64 replied:
Have set defaults before the skips.  Not sure about changing all the testcases to assume the defaults, but I did it anyway to see what you think.

rhansen commented on t/geturl_connectivity.pl.in:
```diff
@@ -13,6 +13,9 @@
     );
     defined($ipv6_socket);
 };
+my $has_curl = eval {
+    qx{ @CURL@ --version 2>/dev/null; } && $? == 0;
+};
```
Why is this wrapped in an `eval`?

dkerr64 replied:
Looks like not necessary.  I have removed it.

rhansen commented on t/geturl_connectivity.pl.in:
```diff
@@ -55,35 +58,52 @@ sub run_httpd {
+    { ipv6_opt => 1, client_ipv => '4', server_ipv => '4', ssl => 1, curl => 0 },
+    { ipv6_opt => 1, client_ipv => '6', server_ipv => '6', ssl => 0, curl => 0 },
+    { ipv6_opt => 1, client_ipv => '6', server_ipv => '6', ssl => 1, curl => 0 },
+    { ipv6_opt => 0, client_ipv => '',  server_ipv => '4', ssl => 0, curl => 1 },
+    { ipv6_opt => 0, client_ipv => '',  server_ipv => '4', ssl => 1, curl => 1 },
+    { ipv6_opt => 0, client_ipv => '4', server_ipv => '4', ssl => 0, curl => 1 },
+    { ipv6_opt => 0, client_ipv => '4', server_ipv => '4', ssl => 1, curl => 1 },
+    { ipv6_opt => 0, client_ipv => '6', server_ipv => '6', ssl => 0, curl => 1 },
+    { ipv6_opt => 0, client_ipv => '6', server_ipv => '6', ssl => 1, curl => 1 },
```
You're missing a couple of cases:

```perl
    { ipv6_opt => 0, client_ipv => '',  server_ipv => '6', ssl => 0, curl => 1 },
    { ipv6_opt => 0, client_ipv => '',  server_ipv => '6', ssl => 1, curl => 1 },
```

dkerr64 replied:
Thanks, not sure how that happened.  Added now.

rhansen commented on t/geturl_ssl.pl:
```diff
@@ -4,6 +4,7 @@
     require HTTP::Request;
     require HTTP::Response;
     require IO::Socket::IP;
+    require IO::Socket::SSL;
```
Add `IO::Socket::SSL` to the optional test modules in `configure.ac`.

dkerr64 replied:
Done

dkerr64 commented on ddclient.in:
```diff
@@ -2087,9 +2096,13 @@ sub geturl {
     if ($use_ssl) {
         # IO::Socket::SSL will load IPv6 support if available on the system.
         load_ssl_support;
+        no strict "subs";
```
Yes, looks like it can.

dkerr64 commented on ddclient.in:
```diff
@@ -2170,6 +2183,163 @@ sub geturl {
+        print($tfh @params);
+    }
+    close($tfh);
+    my $reply = qx{ $system_curl --config '$tmpfile' 2>/dev/null; };
```
done

dkerr64 replied:
Done

dkerr64 commented on ddclient.in:
```diff
@@ -2170,6 +2183,163 @@ sub geturl {
+    return '' if (scalar(@params) == 0); ## no parameters provided
+
+    ## Use system File::Temp to generate filename...
+    use File::Temp;
```
done

dkerr64 commented on ddclient.in:
```diff
@@ -2170,6 +2183,163 @@ sub geturl {
+        print($tfh @params);
+    }
+    close($tfh);
+    my $reply = qx{ $system_curl --config '$tmpfile' 2>/dev/null; };
```
We are catching errors in the return code value and printing  it out ourselves.

dkerr64 commented on ddclient.in:
```diff
@@ -2170,6 +2183,163 @@ sub geturl {
+        print($tfh @params);
+    }
+    close($tfh);
+    my $reply = qx{ $system_curl --config '$tmpfile' 2>/dev/null; };
```
I have added quotemeta() for the curl command and tested it still works (it escapes the forward slash in paths with backslash).  But I also so think we can default to just curl, at least in ddclient.in.

dkerr64 commented on ddclient.in:
```diff
@@ -2170,6 +2183,163 @@ sub geturl {
+
+    ## Use system File::Temp to generate filename...
+    use File::Temp;
+    $tfh = File::Temp->new( TMPDIR => "/tmp/",
```
changed per code review

dkerr64 commented:
Amazing how one simple change request, to add curl, has taken 42 commits (so far) and changed 7 files.  Who would have thought.

dkerr64 commented:
> Amazing how one simple change request, to add curl, has taken 42 commits (so far) and changed 7 files. Who would have thought.

And 196 comments !

dkerr64 commented:
I have added support for `WWW::Curl::Easy` so that this can be compared to using the system curl command and we can discuss the merits of each.

dkerr64 commented:
@rhansen @SuperSandro2000 Gentle nudge on this one.  Can we merge please?

SuperSandro2000 commented on configure.ac:
```diff
@@ -27,6 +27,8 @@ AC_PROG_MKDIR_P
 AC_PATH_PROG([FIND], [find])
 AS_IF([test -z "${FIND}"], [AC_MSG_ERROR(['find' utility not found])])
 
+AC_PATH_PROG([CURL], [curl])
```
If we get a request we could make this dependency optional in the future but for now I would just leave it like it is.

SuperSandro2000 reviewed (approved):
I am going to make a quick test and then merge this. :tada: 

Edit: I did some testing and couldn't find any issues. I am sure we missed something but we are going to fix this as the issues start coming in.

@rhansen If you have any other comments we are going to resolve them in other PRs. I want to get feedback in and see if we broke other providers or if they behave unexpectedly.
